app/blog/posts/good-python-code.mdx

```diff
@@ -0,0 +1,333 @@
+
+I think it's easy to quickly complicate a codebase with premature abstractions without much effort over time - just look at Java. After all, it's a natural reflex as we work towards writing code that is DRY. But that often makes it difficult for you to adapt your code down the line. 
+
+For instance, an easy way to do this initially is to just return a simple dictionary if the returned value is only being used by a single function. 
+
+```
+def extract_attributes(data):
+    new_data = process_data(data)
+    return {"key1": new_data["key1"], "key2": new_data["key2"]}
+
+
+def main():
+    data = pd.read_csv("data.csv")
+    attributes = extract_attributes(data)
+    return attributes
+```
+
+In this example, there's a limited utility to declaring an entire dataclass because it adds additional overhead and complexity to the function. 
+
+```python
+@dataclass
+class Attributes:
+    key1: List[int]
+    key2: List[int]
+
+
+def extract_attributes(data):
+    new_data = process_data(data)
+    return Attributes(key1=new_data["key1"], key2=new_data["key2"])
+
+
+def main():
+    data = pd.read_csv("data.csv")
+    attributes = extract_attributes(data)
+    return attributes
+```
+
+## Decouple your implementation
+
+Another example I like a lot is scoring values. Say we want to calculate the recall for a list of predictions that we've made where we have a single known label as our ground truth and a list of other labels as our model's predictions. We might implement it down the line as 
+
+```
+def calculate_recall(labels,predictions):
+    scores = []
+    for label,preds in zip(labels,predictions):
+        calculate_recall(label,preds)
+    return scores
+```
+
+But what if we'd like to work with other metrics down the line like precision, NDCG or Mean Reciprocal Rank? Wouldn't we then have to declare 4 different functions for this? 
+
+```
+def calculate_recall_predictions(labels, predictions):
+    scores = []
+    for label, preds in zip(labels, predictions):
+        calculate_recall(label, preds)
+    return scores
+
+
+def calculate_ndcg_predictions(labels,predictions):
+    scores = []
+    for label, preds in zip(labels, predictions):
+        calculate_ndcg(label, preds)
+    return scores
+```
+
+A better solution instead is to parameterize the scoring function itself. If you look at the different functions we've defined, they all take in a single label. We're also doing the same thing in each function, which is to score the predictions with respect to a specific output.
+
+This means that we could rewrite this as seen below
+
+```
+def score(labels, predictions, score_fn):
+    return [score_fn(label, pred) for label, pred in zip(labels, predictions)]
+```
+
+In fact, we could even go one step further and just represent all of the metrics that we want as a single dictionary.
+
+```
+SIZES = [3, 10, 15, 25]
+metrics = {"recall": calculate_recall, "mrr": calculate_mrr}
+evals = {}
+
+
+def slice_predictions_at_k(k: int, score):
```

Comment:
```suggestion
def predictions_at_k(score_fn, k:int):
```
the *fn makes it know its a function

k is a property of the s function k so it hould be second

k suggests slice